From e94500eade319d6fae7c7ec6aea8070a8ec5c41c Mon Sep 17 00:00:00 2001
From: ran shalit <rshalit@habana.ai>
Date: Thu, 14 Jul 2022 19:17:52 +0300
Subject: [PATCH] [SW-94540] gaudi3: validate BTL

validate functionality of BTL with the 4 commands.
create empty uboot, preboot binaries.
and remove dependency on ppboot, as we don't
want to compile the file for arm (actually it will
fail compilation anyway)

Change-Id: I7363b2bdb48223e00ab31d9b8ab1505299c53610
Signed-off-by: ran shalit <rshalit@habana.ai>
---

diff --git a/board/habanalabs/common/post-image-spi.sh b/board/habanalabs/common/post-image-spi.sh
index 02915dc..90751c1 100755
--- a/board/habanalabs/common/post-image-spi.sh
+++ b/board/habanalabs/common/post-image-spi.sh
@@ -116,7 +116,15 @@
 }
 
 pushd ${BINARIES_DIR}
-	${HOST_DIR}/usr/bin/mkimage -C none -A arm64 -T script -d ${UBOOT_CMDFILE} boot.scr
+echo sssssssssssssssss
+	touch preboot.bin
+	touch u-boot.bin
+	touch pcie_phy_fw.bin
+
+	mkdir -p ${BUILDROOT_ROOT}/build/linux-custom/
+	touch ${BUILDROOT_ROOT}/build/linux-custom/localversion-habanalabs
+	mkdir -p ${BUILDROOT_ROOT}/build/uboot-custom/
+	touch ${BUILDROOT_ROOT}/build/uboot-custom/localversion-habanalabs
 
 	handle_btl sectors-be.bin ${HL_DEV}-sectors.bin
 	BTL_BINFILE=${HL_DEV}-sectors.bin
@@ -140,12 +148,8 @@
 	dd status=none conv=sync bs=128K count=1 if=/dev/zero | tr "\000" "\377"	 > 128K-0xff.bin
 
 	# Metadata: Pointers
-	prep_image_ptrs ${HL_DEV}-images-pointers.bin.le
-	swap_file ${HL_DEV}-images-pointers.bin.le ${HL_DEV}-images-pointers.bin.be
-	dd status=none conv=sync bs=64K count=1 if=64K-0xff.bin				 > ${HL_DEV}-images-pointers.bin.be.spi
-	dd status=none conv=sync,notrunc bs=4 if=${HL_DEV}-images-pointers.bin.be of=${HL_DEV}-images-pointers.bin.be.spi
 
-	create_spi_image ${BTL_BINFILE} ${PREBOOT_BINFILE} ${HL_DEV}-images-pointers.bin.be.spi ${HL_DEV}-spi-flash.bin.nocrc
+	create_spi_image ${BTL_BINFILE} ${PREBOOT_BINFILE} u-boot.bin	 ${HL_DEV}-spi-flash.bin.nocrc
 
 	# Prepare checksum
 	CKSUM=`${SHA_CHKSUM} ${HL_DEV}-spi-flash.bin.nocrc | cut -d' ' -f1`
diff --git a/board/habanalabs/gaudi3/post-image-spi.sh b/board/habanalabs/gaudi3/post-image-spi.sh
index fdc00d8..49a3b7d 100755
--- a/board/habanalabs/gaudi3/post-image-spi.sh
+++ b/board/habanalabs/gaudi3/post-image-spi.sh
@@ -1,3 +1,71 @@
 #!/usr/bin/env bash
 
-# TODO fill with content
+set -e  	# stop on any error in this script
+set -u 		# stop on any reference to undefine variable
+set -o pipefail # stop on any error in pipeline
+
+function handle_btl {
+	local INF=$1
+	local OUTF=$2
+
+	for s in s?.bin; do
+		# Align to 4B
+		dd status=none conv=sync bs=4 if=${s} of=${s}.le
+	done
+
+	# Align to 64KB
+	dd status=none conv=sync bs=64K if=${INF} of=${INF}.tmp
+	swap_file ${INF}.tmp ${OUTF}
+}
+
+PREBOOT_OFFSET=0x80000
+PREBOOT2_OFFSET=0xA0000
+if [ ${HL_TARGET} == "onix" ]; then
+	PREBOOT_TEXT_BASE=0x8200005000
+else
+	PREBOOT_TEXT_BASE=0x7FFD005000
+fi
+
+SHA_CHKSUM=sha384sum
+
+function hl_compress_bin {
+	#compression not implemented in gaudi2
+	return
+}
+
+function create_spi_image {
+	local BTL_BIN=$1
+	local PREBOOT_BIN=$2
+	local IMG_PTR_BIN=$3
+	local OUTF=$4
+
+	# [0-5] BTL (pri,sec,prstn1,prstn2,pcie1,pcie2)
+	dd status=none conv=sync bs=64K		if=${BTL_BIN}		 > ${OUTF}
+	# [6] Image pointers ARM
+	dd status=none conv=sync bs=64K		if=${IMG_PTR_BIN}	>> ${OUTF}
+	# [7] U-Boot ENV
+	dd status=none conv=sync bs=64K count=1 if=64K-0xff.bin		>> ${OUTF}
+	# [8-11] Preboot1
+	dd status=none conv=sync bs=256K	if=${PREBOOT_BINFILE}	>> ${OUTF}
+	# [12-15] Preboot2
+	dd status=none conv=sync bs=256K	if=${PREBOOT_BINFILE}	>> ${OUTF}
+}
+
+function hl_chksum_add {
+	# As a part of image descriptor version 3, an additional checksum field
+	# was added to support sha384, just after the sha256 field. So this
+	# function populates a valid checksum for the relavent algorithm and
+	# pads the irrelavant one. For projects using sha256  the first field
+	# is populated with a valid checksum while the second field is padded.
+	# For projects using sha384 it is vice versa.
+	local pad_dwords=8
+	local checksum=$1
+
+	# Pad the sha256 checksum field
+	hl_chksum_padding ${pad_dwords}
+	hl_chksum_padding ${pad_dwords}
+
+	# Update the sha384 checksum field
+	echo $checksum
+	echo $checksum
+}
diff --git a/configs/gaudi3_arm_defconfig b/configs/gaudi3_arm_defconfig
index 9303fd2..d44d739 100644
--- a/configs/gaudi3_arm_defconfig
+++ b/configs/gaudi3_arm_defconfig
@@ -8,7 +8,6 @@
 BR2_TARGET_GENERIC_HOSTNAME="$(HL_DEV)"
 BR2_TARGET_GENERIC_ISSUE="Welcome to Gaudi3"
 BR2_ROOTFS_OVERLAY="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/common/rootfs_overlay"
-BR2_ROOTFS_POST_BUILD_SCRIPT="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/common/post-build.sh"
-# BR2_POST_IMAGE_FW_LOADER_ITS_TEMPLATE="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/$(HL_DEV)/agent-fw_loader.its"
-# BR2_POST_IMAGE_AGENTS_ITS_TEMPLATE="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/$(HL_DEV)/agent.its"
-BR2_PACKAGE_ZEPHYR_SAMPLE_TEST=y
\ No newline at end of file
+BR2_PACKAGE_ZEPHYR_SAMPLE_TEST=y
+# BR2_ROOTFS_POST_BUILD_SCRIPT="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/common/post-build.sh"
+BR2_ROOTFS_POST_IMAGE_SCRIPT="$(BR2_EXTERNAL_HABANALABS_PATH)/board/habanalabs/common/post-image-spi.sh"
diff --git a/package/btl/btl.mk b/package/btl/btl.mk
index 06f867b..ce045e0 100644
--- a/package/btl/btl.mk
+++ b/package/btl/btl.mk
@@ -2,7 +2,6 @@
 BTL_SITE = btl
 BTL_LICENSE = unclear
 
-BTL_DEPENDENCIES = preboot
 
 define BTL_BUILD_CMDS
 	$(TARGET_MAKE_ENV) $(MAKE) -C $(@D) all
